File: 兴趣班入门百练讲义/NQ100/NQ080/NQ080_improved.cpp
#include <iostream>
#include <cstring>
using namespace std;
bool p[100];
bool used[100];
int a[100];
int n, now = 0;
void dfs(int deep) {
    if (deep == n + 1 && p[a[1] + a[n]]) {
        for (int i = 1; i <= n; i++) {
            cout << a[i];
            if (i != n) cout << " ";
        }
        cout << endl;
        return;
    }
    for (int i = 2; i <= n; i++) {
        if (!used[i] && p[i + a[deep - 1]]) {
            used[i] = true;
            a[deep] = i;
            dfs(deep + 1);  // 递归搜索下一个位置
            used[i] = false;  // 回溯
        }
    }
}
int main() {
    // 预处理质数表
    memset(p, false, sizeof(p));
    p[2] = p[3] = p[5] = p[7] = p[11] = true;
    p[13] = p[17] = p[19] = p[23] = p[29] = true;
    p[31] = p[37] = true;
    bool first_case = true;
    while (cin >> n) {
        a[1] = 1;  // 第一个元素固定为1
        if (!first_case) {
            cout << endl;
        }
        first_case = false;
        cout << "Case " << ++now << ":" << endl;
        memset(used, false, sizeof(used));
        dfs(2);  // 从第二个位置开始DFS
    }
    return 0;
}

File: source/algorithm/NQ080/NQ080_improved.py
# 质数标记数组，p[i]为True表示i是质数
p = [False] * 100
# 初始化质数表
p[2] = p[3] = p[5] = p[7] = p[11] = True
p[13] = p[17] = p[19] = p[23] = p[29] = True
p[31] = p[37] = True

now = 0

def dfs(deep, n, a, used):
    if deep == n and p[a[0] + a[-1]]:
        print(' '.join(map(str, a)))
        return
    for i in range(2, n + 1):
        if not used[i] and p[i + a[-1]]:
            used[i] = True
            a.append(i)
            dfs(deep + 1, n, a, used)  # 递归搜索下一个位置
            a.pop()  # 回溯
            used[i] = False  # 回溯

def main():
    global now
    first_case = True
    import sys
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        n = int(line)
        if not first_case:
            print()
        first_case = False
        now += 1
        print("Case {}:".format(now))
        used = [False] * (n + 1)
        a = [1]  # 第一个元素固定为1
        used[1] = True
        dfs(1, n, a, used)  # 从第二个位置开始DFS

if __name__ == "__main__":
    main()
